Support for non-flattened array multiplication

cako opened the pull request:
Follow-up to #348.

This draft PR implements a basic support for non-flat vector, in hopes of spurring further discussion on design choices for this feature.

The key part of the implementation relies in [`LinearOperator.dot`](https://github.com/cako/pylops/blob/9ecd38010c7a2c7a9b8a0a230f563ed225451e4d/pylops/LinearOperator.py#L233) identifying whether `dims` and `dimsd` are present. If this is true, it flattens the array and passes it onto `LinearOperator.matvec`. Upon the return of a (flattened) vector, it then reshapes to `dimsd` any array that came in as a `dims`-shaped operator. In this way, matrix-matrix multiplications remain unchanged and do no support ndarrays.

An example of expected behavior using the Symmetrize operator can be found in [FlattenDims.ipynb](https://github.com/cako/pylops_notebooks/blob/master/developement/FlattenDims.ipynb), summarized below:

|     dims   |    dimsd   | expected dimsd | correct? |
--------------|--------------|----------------------|-------------|
|   (20, 20) |   (30, 20) |     (30, 20)   | ✓|
|     (400,) |    error   |       error    | ✓|
|(2, 10, 20) |    error   |       error    | ✓|
|    (2, 10) |    (3, 10) |      (3, 10)   | ✓|
| (2, 10, 1) |    error   |       error    | ✓|
|      (20,) |      (30,) |        (30,)   | ✓|
|    (20, 1) |    (30, 1) |      (30, 1)   | ✓|

There is an argument for using matmat for `(2, 10, X)` instead of erroring, but for the sake of simplicity this was not implemented.

In addition to the changed to `LinearOperator`, I am also providing a decorator to simplify the reshape/flatten pattern, exemplified in this PR by Symmetrize:
 https://github.com/PyLops/pylops/blob/9ecd38010c7a2c7a9b8a0a230f563ed225451e4d/pylops/basicoperators/Symmetrize.py#L75-L93

### Next steps
If we want to proceed with this, there are a few things to sort out. First, we need to decide which shapes are accepted, and which are not. The above table is only a suggestion.

Once that is defined, we need to make dims/dimsd "first-class" attributes which get passed onto transpose/adjoint, multiplications, etc.

mrava87 commented on pylops/LinearOperator.py:
```diff
@@ -253,12 +253,32 @@ def dot(self, x):
+            dims = getattr(self, "dims", None)
+            dimsd = getattr(self, "dimsd", None)
+
+            reshape_out = False
```
Can you please add a comment here, explaining what this piece of code does... that it will reshape the output back into a ndarray of size dimsd when the input had shape dims

mrava87 commented on pylops/LinearOperator.py:
```diff
@@ -253,12 +253,32 @@ def dot(self, x):
         else:
-            if x.ndim == 1:  # or x.ndim == 2 and x.shape[1] == 1:
-                return self.matvec(x)
+            dims = getattr(self, "dims", None)
```
ultimately we should avoid this check by making sure we always have dims and dimsd (same as shape and dtype)

mrava87 replied:
I checked

```
In [7]: %timeit getattr(Mop, "dims", None)
60.5 ns ± 0.315 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [8]: %timeit Mop.dims
37.2 ns ± 1.3 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

though I am not sure if we care as it is an irrelevant time difference...

mrava87 commented on pylops/avo/avo.py:
```diff
@@ -627,7 +627,7 @@ def _matvec(self, x):
             )
         y = self.ncp.sum(self.G * x[:, :, self.ncp.newaxis], axis=1)
-        return y
+        return y.ravel()
```
We should consider making this operator consistent with others, using `dims` and `dimsd` :)

Also, why is this needed now. In the current dev:

```
(pylops.avo.AVOLinearModelling(np.ones(5),nt0=100) * np.ones(300)).shape
Out[21]: (500,)
```
and
```
(pylops.avo.AVOLinearModelling(np.ones(5),nt0=100, spatdims=20) * np.ones(300*20)).shape
Out[23]: (10000,)
```

mrava87 commented on pylops/utils/_internal.py:
```diff
@@ -1,6 +1,49 @@
+
+        return wrapper
+
+    return decorator
```
this looks very good, probably we should add this routine to the documentation in the PyLops utilities (it is true it is only for developers but it is something a bit more `exposed` than things like `_value_or_list_like_to_array` and it may start to be present in most operators.

What do you think?

cako commented:
I reorganized this entirely based on the idea of using properties in LinearOperators introduced in #365. The current version works like this:

1. Every LOp has `dims` and `dimsd`. They are arbitrary tuples which must be compatible with shape.
2. Suppose that LOp.dims = (5, 10) and LOp.dimsd = (9, 10). The following rules shapes are obtained on LOp @ x

   |Reference|     x.shape   |    (LOp @ x).shape   | Note | 
   ---|--------------|--------------|-----|
   V0|      (50,) |      (90,) | Standard vector multiplication
   V1|    (5, 10) |    (9, 10) | "Vector" of size (5, 10)
   M0|    (50, 1) |    (90, 1) | Standard one-column matrix multiplication
   M1| (5, 10, 1) |    (9, 10, 1)   | "One-column matrix" of "vector" (5, 10)
   M2|   (50, 20) |   (90, 20) | Standard matrix multiplication
   M3|(5, 10, 20) |    (9, 10, 20)   | "Matrix" of 20 x (5, 10)
   ||     (1000,) |    error   | Can be reshaped to (50, 20) but size is ambigous
   || any other kind of shape | error |
3. SciPy solvers will struggle with V1, M1 and M3. These cases can be coerced to error if LOp.flattened is set to True. Any operator with flattened = True will have legacy behavior. I'm happy to hear about other ways on how we can do this.

In addition, I have moved and refactored the ``reshape`` decorator. It now relies primarily on the name of the function being decorated to find out which dimensions to reshape by, and also includes an optional swapaxis. See the new Symmetrize operator for examples.

cako commented:
Moved to #367
